File: src/updateDistance.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <cpp_solver/msg/edge_distance.hpp>
#include <memory>
#include <vector>
#include <string>
#include <sstream>
#include <iostream>
#include <map>
#include <set>

#include "cpp_solver/RequestGraph.h"
#include "cpp_solver/EdgeDistance.h"

#include "AStar/AStar2.h"

std::map<int, std::pair<float, float>> verticesPosition;
std::vector<int> vertices;
std::set<std::pair<int, int>> edgeSet;
bool getPriorMap;

std::vector<std::pair<int, int>> dir;
std::set<std::pair<int, int>> updated_edges;

std::vector<double> robotInitPosition;
rclcpp::Publisher<cpp_solver::msg::EdgeDistance>::SharedPtr pubDistance;
std::shared_ptr<rclcpp::Node> node;

void occupancyGridCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
    RCLCPP_INFO(node->get_logger(), "Received an OccupancyGrid message!");
    RCLCPP_INFO(node->get_logger(), "Width: %d, Height: %d, Size: %d", msg->info.width, msg->info.height, msg->data.size());
    geometry_msgs::msg::Pose originPoint = msg->info.origin;
    RCLCPP_INFO(node->get_logger(), "Map origin x: %.3f, y: %.3f", originPoint.position.x, originPoint.position.y);
    
    if(!getPriorMap){
        RCLCPP_WARN(node->get_logger(), "Do not get prior map yet.");
        return;
    }
    // A* path finder
    float resolution = msg->info.resolution;
    int width = msg->info.width, height = msg->info.height;
    std::unique_ptr<int8_t[]> map_arr(new int8_t[msg->data.size()]);
    for(unsigned int i = 0; i < msg->data.size(); i++){
        // Value: -1 for unknown, 0 for free, 100 for occupied
        map_arr[i] = static_cast<int8_t>(msg->data[i]);
    }

    AStar::PathFinder astar_finder;
    astar_finder.setWorldData(static_cast<unsigned int>(msg->info.width), 
                            static_cast<unsigned int>(msg->info.height), 
                            map_arr.get());
    // Get vertices coordindate index in map, and only remain free vertices
    std::vector<int> free_vertices;
    std::vector<std::pair<int, int>> free_vertices_index;
    for(int i = 0; i < verticesPosition.size(); i++){
        int v = vertices[i];
        int v_map_x = std::round((verticesPosition[v].first - originPoint.position.x) / resolution);
        int v_map_y = std::round((verticesPosition[v].second - originPoint.position.y) / resolution);
        
        if(v_map_x >= width || v_map_x < 0 || v_map_y >= height || v_map_y <= 0)
            continue;
        int index = v_map_x + std::max(0, v_map_y-1) * width;

        if(map_arr[index] < 0 || map_arr[index] > 60)
            continue;
        // Check whether neighbors are free
        int count_free_neighbor = 0;
        for(int j = 0; j < dir.size(); j++){
            int neighbor_x = v_map_x + dir[j].first;
            int neighbor_y = v_map_y + dir[j].second;
            if(neighbor_x >= width || neighbor_x < 0 || neighbor_y >= height || neighbor_y <= 0)
                continue;

            int neighbor_index = neighbor_x + std::max(0, neighbor_y-1) * width;
            if(map_arr[neighbor_index] == 0){
                count_free_neighbor++;
            }
            // if(v == 8)
            //     RCLCPP_WARN("cell value: %d", map_arr[neighbor_index]);
        }
        if(count_free_neighbor == 8){
            free_vertices.push_back(v);
            free_vertices_index.push_back(std::make_pair(v_map_x, v_map_y));
        }
        // if(v == 8)
        //     RCLCPP_WARN("free cell number: %d", count_free_neighbor);
    }
    RCLCPP_INFO(node->get_logger(), "Free vertices size: %d", free_vertices.size());

    if(free_vertices.size() > 1){
        // Update edge weight and publish
        AStar::CoordinateList astar_path;
        for(int k = 0; k < free_vertices.size(); k++){
            for(int p = 0; p < k; p++){
                std::pair<int, int> candidate = std::make_pair(free_vertices[k], free_vertices[p]);
                std::pair<int, int> candidate_reverse = std::make_pair(free_vertices[p], free_vertices[k]);
                if(edgeSet.find(candidate) != edgeSet.end() || updated_edges.find(candidate) != updated_edges.end()
                   || edgeSet.find(candidate_reverse) != edgeSet.end() || updated_edges.find(candidate_reverse) != updated_edges.end()){
                    continue;
                }
                astar_path = astar_finder.findPath({free_vertices_index[k].first, free_vertices_index[k].second}, 
                                                {free_vertices_index[p].first, free_vertices_index[p].second});
                if(astar_path.size() == 0)
                    continue;
                // More accurate distance estimation
                int prev_x = -1, prev_y = -1;
                double astarDist = 0;
                for(auto& point: astar_path){
                    if(prev_x >= 0){
                        if(std::abs(point.x - prev_x) + std::abs(point.y - prev_y) > 1){
                            astarDist += 1.414 * resolution;
                        }else{
                            astarDist += resolution;
                        }
                    }
                    prev_x = point.x;
                    prev_y = point.y;
                }
                // FIXME: astarDist should be less than distance on the prior map
                cpp_solver::msg::EdgeDistance edge_distance;
                edge_distance.vertex1 = free_vertices[k];
                edge_distance.vertex2 = free_vertices[p];
                edge_distance.distance = astarDist;
                pubDistance->publish(edge_distance);
                updated_edges.insert(candidate);           
            }
        }
    }
    RCLCPP_INFO(node->get_logger(), "%d edge distance updated", updated_edges.size());
    return;
}

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    node = rclcpp::Node::make_shared("updateDistance");

    dir.push_back(std::make_pair(0, 1));
    dir.push_back(std::make_pair(0, -1));
    dir.push_back(std::make_pair(1, 0));
    dir.push_back(std::make_pair(-1, 0));
    dir.push_back(std::make_pair(1, 1));
    dir.push_back(std::make_pair(1, -1));
    dir.push_back(std::make_pair(-1, 1));
    dir.push_back(std::make_pair(-1, -1));

    getPriorMap = false;
	RCLCPP_INFO(node->get_logger(), "Waiting for prior_graph_service ...");

    //auto clientPriorGraph = node->create_client<cpp_solver::srv::RequestGraph>("prior_graph_service");

	// Waiting for service is not implemented yet
	//bool priorGraphServiceAvailable = rclcpp::service::waitForService("prior_graph_service", rclcpp::Duration(30.0));
    //if (!priorGraphServiceAvailable)
    //{
    //    RCLCPP_ERROR(node->get_logger(), "prior_graph_service not available.");
    //}
	//else{
		//rclcpp::Client<cpp_solver::srv::RequestGraph>::SharedPtr clientPriorGraph = node->create_client<cpp_solver::srv::RequestGraph>("prior_graph_service");
		//auto request = std::make_shared<cpp_solver::srv::RequestGraph::Request>();
		//auto result = clientPriorGraph->async_send_request(request);
		//if (rclcpp::spin_until_complete(node,
        //                      std::chrono::duration_cast<std::chrono::seconds>(std::chrono::seconds(30))) ==
        //rclcpp::FutureReturnCode::TIMEOUT)
        //{
        //    RCLCPP_ERROR(node->get_logger(), "prior_graph_service not available.");
		//}
		//else
		//{	
		//	//cpp_solver::RequestGraph::Response res2 = clientMessage2.response;
		//	vertices = result.get()->response.vertices;
		//	std::vector<float> x_coords = result.get()->response.x_coords;
		//	std::vector<float> y_coords = result.get()->response.y_coords;
		//	std::vector<int32_t> edges_start = result.get()->response.edges_start;
		//	std::vector<int32_t> edges_end = result.get()->response.edges_end;
		//	RCLCPP_INFO(node->get_logger(), "Receive response of service prior_graph_service.");
        //    RCLCPP_INFO(node->get_logger(), "Prior map has %d vertices, %d edges", vertices.size(), edges_start.size());
        //    for(int i = 0; i < vertices.size(); i++){
        //        verticesPosition[vertices[i]] = std::make_pair(x_coords[i], y_coords[i]);
        //    }
        //    for(int i = 0; i < edges_start.size(); i++){
        //        edgeSet.insert(std::make_pair(edges_start[i], edges_end[i]));
        //    }
        //    getPriorMap = true;
		//}
		//else
		//{
		//	RCLCPP_ERROR(node->get_logger(), "Failed to call service: prior_graph_service");
		//}
	//}

    // Read the robot_init_position parameter as a string from the parameter server
    std::string robotInitPositionStr;
    node->declare_parameter("robot_start_position", std::string("0.0 0.0 0.0"));
    if (!node->get_parameter("robot_start_position", robotInitPositionStr)) {
        RCLCPP_ERROR(node->get_logger(), "Failed to read robot_init_position parameter.");
        return 1;
    }

    // Parse the string to get three double values
    std::istringstream iss(robotInitPositionStr);
    double value;
    while (iss >> value) {
        robotInitPosition.push_back(value);
        // Check if there are exactly three values in the string
        if (robotInitPosition.size() == 3) {
            break;
        }
    }

    // Check if there are exactly three values in the string
    if (robotInitPosition.size() != 3) {
        RCLCPP_ERROR(node->get_logger(), "Invalid robot_init_position parameter format. Expected three double values separated by spaces.");
        return 1;
    }
    RCLCPP_INFO(node->get_logger(), "Get robot position: %.2f, %.2f", robotInitPosition[0], robotInitPosition[1]);

    // Create subscriber and publisher
    auto sub = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
        "/map", rclcpp::QoS(1), occupancyGridCallback);

    pubDistance = node->create_publisher<cpp_solver::msg::EdgeDistance>("/edge_distance", 10);

    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}

File: cpp_solver/update_distance.py

#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
import numpy as np
import math
from typing import List, Tuple, Dict, Set, Optional

from nav_msgs.msg import OccupancyGrid
from cpp_solver.msg import EdgeDistance
from cpp_solver.srv import RequestGraph

class AStar:
    def __init__(self):
        self.directions = [(-1, -1), (0, -1), (1, -1), (-1, 0), (1, 0), (-1, 1), (0, 1), (1, 1)]
        self.direction_costs = [14, 10, 14, 10, 10, 14, 10, 14]
    
    def set_world_data(self, width: int, height: int, data: List[int]):
        self.width = width
        self.height = height
        self.data = data
    
    def find_path(self, start: Tuple[int, int], goal: Tuple[int, int]) -> List[Tuple[int, int]]:
        """Find path using A* algorithm"""
        if self.is_collision(start) or self.is_collision(goal):
            return []
        
        open_set = []
        closed_set = set()
        g_score = {}
        f_score = {}
        came_from = {}
        
        g_score[start] = 0
        f_score[start] = self.heuristic(start, goal)
        open_set.append((f_score[start], start))
        
        while open_set:
            open_set.sort()
            current_f, current = open_set.pop(0)
            
            if current == goal:
                return self.reconstruct_path(came_from, current)
            
            closed_set.add(current)
            
            for i, direction in enumerate(self.directions):
                neighbor = (current[0] + direction[0], current[1] + direction[1])
                
                if self.is_collision(neighbor) or neighbor in closed_set:
                    continue
                
                tentative_g = g_score[current] + self.direction_costs[i]
                
                if neighbor not in g_score or tentative_g < g_score[neighbor]:
                    came_from[neighbor] = current
                    g_score[neighbor] = tentative_g
                    f_score[neighbor] = tentative_g + self.heuristic(neighbor, goal)
                    
                    if (f_score[neighbor], neighbor) not in open_set:
                        open_set.append((f_score[neighbor], neighbor))
        
        return []
    
    def is_collision(self, pos: Tuple[int, int]) -> bool:
        """Check if position is collision"""
        x, y = pos
        if x < 0 or x >= self.width or y < 0 or y >= self.height:
            return True
        
        index = y * self.width + x
        if index >= len(self.data):
            return True
        
        return self.data[index] >= 60 or self.data[index] < 0
    
    def heuristic(self, a: Tuple[int, int], b: Tuple[int, int]) -> float:
        """Manhattan distance heuristic"""
        return abs(a[0] - b[0]) + abs(a[1] - b[1])
    
    def reconstruct_path(self, came_from: Dict, current: Tuple[int, int]) -> List[Tuple[int, int]]:
        """Reconstruct path from came_from dictionary"""
        path = [current]
        while current in came_from:
            current = came_from[current]
            path.append(current)
        return path[::-1]

class UpdateDistance(Node):
    def __init__(self):
        super().__init__('update_distance')
        
        # Parameters
        self.declare_parameter('robot_start_position', '0.0 0.0 0.0')
        robot_pos_str = self.get_parameter('robot_start_position').value
        self.robot_init_position = [float(x) for x in robot_pos_str.split()]
        
        # Initialize variables
        self.vertices_position = {}
        self.vertices = []
        self.edge_set = set()
        self.get_prior_map = False
        self.updated_edges = set()
        
        # Directions for checking neighbors
        self.directions = [
            (0, 1), (0, -1), (1, 0), (-1, 0),
            (1, 1), (1, -1), (-1, 1), (-1, -1)
        ]
        
        # QoS profile
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.RELIABLE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10
        )
        
        # Publishers
        self.pub_distance = self.create_publisher(EdgeDistance, '/edge_distance', qos_profile)
        
        # Subscribers
        self.sub_map = self.create_subscription(
            OccupancyGrid, '/map', self.occupancy_grid_callback, qos_profile)
        
        # Service clients
        self.client_prior_graph = self.create_client(RequestGraph, 'prior_graph_service')
        
        # Initialize A* pathfinder
        self.astar_finder = AStar()
        
        # Request prior graph
        self.request_prior_graph()
        
        self.get_logger().info(f"Get robot position: {self.robot_init_position[0]:.2f}, {self.robot_init_position[1]:.2f}")
    
    def request_prior_graph(self):
        """Request prior graph from service"""
        self.get_logger().info("Waiting for prior_graph_service...")
        
        if not self.client_prior_graph.wait_for_service(timeout_sec=30.0):
            self.get_logger().error("prior_graph_service not available.")
            return
        
        request = RequestGraph.Request()
        future = self.client_prior_graph.call_async(request)
        
        def response_callback(future):
            try:
                response = future.result()
                self.vertices = response.vertices
                x_coords = response.x_coords
                y_coords = response.y_coords
                edges_start = response.edges_start
                edges_end = response.edges_end
                
                self.get_logger().info("Receive response of service prior_graph_service.")
                self.get_logger().info(f"Prior map has {len(self.vertices)} vertices, {len(edges_start)} edges")
                
                # Store vertex positions
                for i, vertex_id in enumerate(self.vertices):
                    self.vertices_position[vertex_id] = (x_coords[i], y_coords[i])
                
                # Store edges
                for i in range(len(edges_start)):
                    self.edge_set.add((edges_start[i], edges_end[i]))
                
                self.get_prior_map = True
                
            except Exception as e:
                self.get_logger().error(f"Failed to call service: prior_graph_service - {e}")
        
        future.add_done_callback(response_callback)
    
    def occupancy_grid_callback(self, msg: OccupancyGrid):
        """Process occupancy grid and update edge distances"""
        self.get_logger().info("Received an OccupancyGrid message!")
        self.get_logger().info(f"Width: {msg.info.width}, Height: {msg.info.height}, Size: {len(msg.data)}")
        
        origin_point = msg.info.origin
        self.get_logger().info(f"Map origin x: {origin_point.position.x:.3f}, y: {origin_point.position.y:.3f}")
        
        if not self.get_prior_map:
            self.get_logger().warn("Do not get prior map yet.")
            return
        
        # Setup A* pathfinder
        resolution = msg.info.resolution
        width = msg.info.width
        height = msg.info.height
        
        # Convert data to proper format
        map_data = list(msg.data)
        self.astar_finder.set_world_data(width, height, map_data)
        
        # Find free vertices
        free_vertices = []
        free_vertices_index = []
        
        for vertex_id in self.vertices:
            if vertex_id not in self.vertices_position:
                continue
            
            vertex_pos = self.vertices_position[vertex_id]
            v_map_x = round((vertex_pos[0] - origin_point.position.x) / resolution)
            v_map_y = round((vertex_pos[1] - origin_point.position.y) / resolution)
            
            # Check bounds
            if v_map_x >= width or v_map_x < 0 or v_map_y >= height or v_map_y <= 0:
                continue
            
            # Check if vertex is free
            index = v_map_x + max(0, v_map_y - 1) * width
            if index >= len(map_data) or map_data[index] < 0 or map_data[index] > 60:
                continue
            
            # Check if all neighbors are free
            count_free_neighbor = 0
            for dx, dy in self.directions:
                neighbor_x = v_map_x + dx
                neighbor_y = v_map_y + dy
                
                if neighbor_x >= width or neighbor_x < 0 or neighbor_y >= height or neighbor_y <= 0:
                    continue
                
                neighbor_index = neighbor_x + max(0, neighbor_y - 1) * width
                if neighbor_index < len(map_data) and map_data[neighbor_index] == 0:
                    count_free_neighbor += 1
            
            if count_free_neighbor == 8:
                free_vertices.append(vertex_id)
                free_vertices_index.append((v_map_x, v_map_y))
        
        self.get_logger().info(f"Free vertices size: {len(free_vertices)}")
        
        if len(free_vertices) <= 1:
            return
        
        # Update edge weights
        updated_count = 0
        for i in range(len(free_vertices)):
            for j in range(i):
                v1, v2 = free_vertices[i], free_vertices[j]
                candidate = (v1, v2)
                candidate_reverse = (v2, v1)
                
                # Skip if edge already exists or already updated
                if (candidate in self.edge_set or candidate_reverse in self.edge_set or
                    candidate in self.updated_edges or candidate_reverse in self.updated_edges):
                    continue
                
                # Find A* path
                start_pos = free_vertices_index[i]
                goal_pos = free_vertices_index[j]
                astar_path = self.astar_finder.find_path(start_pos, goal_pos)
                
                if not astar_path:
                    continue
                
                # Calculate accurate distance
                astar_dist = 0.0
                prev_x, prev_y = -1, -1
                
                for point in astar_path:
                    if prev_x >= 0:
                        if abs(point[0] - prev_x) + abs(point[1] - prev_y) > 1:
                            astar_dist += 1.414 * resolution
                        else:
                            astar_dist += resolution
                    prev_x, prev_y = point
                
                # Publish edge distance
                edge_distance = EdgeDistance()
                edge_distance.vertex1 = v1
                edge_distance.vertex2 = v2
                edge_distance.distance = astar_dist
                self.pub_distance.publish(edge_distance)
                
                self.updated_edges.add(candidate)
                updated_count += 1
        
        self.get_logger().info(f"{updated_count} edge distances updated")

def main(args=None):
    rclpy.init(args=args)
    node = UpdateDistance()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
